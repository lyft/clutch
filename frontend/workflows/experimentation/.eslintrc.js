const baseConfig = require("@clutch-sh/tools/.eslintrc.js");

module.exports = {
  ...baseConfig,
  overrides: [
    ...baseConfig.overrides,
    {
      files: ["**/*.test.*", "**/*.mjs"],
      rules: {
        "import/no-extraneous-dependencies": ["off"],
      },
    },
  ],
};
