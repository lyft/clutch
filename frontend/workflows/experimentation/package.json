{
  "name": "@clutch-sh/experimentation",
  "version": "1.0.0-beta",
  "description": "Clutch Experimentation Workflows",
  "license": "Apache-2.0",
  "author": "clutch@lyft.com",
  "main": "dist/index.js",
  "scripts": {
    "build": "yarn clean && yarn compile",
    "clean": "rm -rf ./dist && rm -f tsconfig.tsbuildinfo",
    "compile": "tsc -b",
    "compile:dev": "node build.mjs $PWD",
    "compile:watch": "node build.mjs $PWD -w",
    "lint": "eslint --ext .js,.jsx,.ts,.tsx .",
    "lint:fix": "yarn lint --fix",
    "publishBeta": "../../../tools/publish-frontend.sh experimentation",
    "test": "jest --passWithNoTests",
    "test:coverage": "yarn test --collect-coverage",
    "test:watch": "yarn test --watch"
  },
  "dependencies": {
    "@clutch-sh/api": "^1.0.0-beta",
    "@clutch-sh/core": "^1.0.0-beta",
    "@clutch-sh/data-layout": "^1.0.0-beta",
    "@emotion/styled": "^11.0.0",
    "@material-ui/core": "^4.11.4",
    "history": "^5.0.0",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-is": "^17.0.2",
    "yup": "^0.32.8"
  },
  "devDependencies": {
    "@clutch-sh/tools": "^1.0.0-beta"
  },
  "engines": {
    "node": ">=14.0.0"
  }
}
