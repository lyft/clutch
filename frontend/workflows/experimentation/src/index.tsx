import type { WorkflowConfiguration } from "@clutch-sh/core";

import FormFields from "./core/form-fields";
import PageLayout from "./core/page-layout";
import ListExperiments from "./list-experiments";
import ViewExperimentRun from "./view-experiment-run";

const register = (): WorkflowConfiguration => {
  return {
    developer: {
      name: "Lyft",
      contactUrl: "mailto:hello@clutch.sh",
    },
    path: "experimentation",
    group: "Chaos Experimentation",
    displayName: "Chaos Experimentation",
    routes: {
      listExperiments: {
        path: "list",
        displayName: "Manage Experiments",
        description: "Manage experiments.",
        component: ListExperiments,
      },
      viewExperimentRun: {
        path: "run/:runID",
        displayName: "View Experiment Run",
        description: "View Experiment Run.",
        hideNav: true,
        component: ViewExperimentRun,
      },
    },
  };
};

export default register;

export { PageLayout, FormFields };
