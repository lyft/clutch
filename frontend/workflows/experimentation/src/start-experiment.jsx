import React from "react";
import {
  ButtonGroup,
  client,
  Confirmation,
  MetadataTable,
  useWizardContext,
} from "@clutch-sh/core";
import { useDataLayout } from "@clutch-sh/data-layout";
import { Wizard, WizardStep } from "@clutch-sh/wizard";
import * as yup from "yup";

const ClusterPairTargetDetails = () => {
  const { onSubmit } = useWizardContext();
  const clusterPairData = useDataLayout("clusterPairTargetData");
  const clusterPair = clusterPairData.displayValue();

  return (
    <WizardStep error={clusterPairData.error}>
      <MetadataTable
        onUpdate={(key, value) => clusterPairData.updateData(key, value)}
        data={[
          {
            name: "Downstream Cluster",
            value: clusterPair.downstreamCluster,
            input: {
              key: "downstreamCluster",
              validation: yup.string().required(),
            },
          },
          {
            name: "Upstream Cluster",
            value: clusterPair.upstreamCluster,
            input: {
              key: "upstreamCluster",
              validation: yup.string().required(),
            },
          },
        ]}
      />
      <ButtonGroup
        buttons={[
          {
            text: "Next",
            onClick: onSubmit,
          },
        ]}
      />
    </WizardStep>
  );
};

const AbortExperimentDetails = () => {
  const { onSubmit, onBack } = useWizardContext();
  const abortExperimentData = useDataLayout("abortExperimentData");
  const abortExperiment = abortExperimentData.value;

  return (
    <WizardStep error={abortExperimentData.error}>
      <MetadataTable
        onUpdate={(key, value) => abortExperimentData.updateData(key, value)}
        data={[
          {
            name: "Percent",
            value: abortExperiment.percent,
            input: {
              type: "number",
              key: "percent",
              validation: yup.number().integer().moreThan(-1).lessThan(101),
            },
          },
          {
            name: "HTTP Status",
            value: abortExperiment.httpStatus,
            input: {
              type: "number",
              key: "httpStatus",
              validation: yup.number().integer().moreThan(99).lessThan(600),
            },
          },
        ]}
      />
      <ButtonGroup
        buttons={[
          {
            text: "Back",
            onClick: onBack,
          },
          {
            text: "Next",
            onClick: onSubmit,
            destructive: true,
          },
        ]}
      />
    </WizardStep>
  );
};

const LatencyExperimentDetails = () => {
  const { onSubmit, onBack } = useWizardContext();
  const latencyExperimentData = useDataLayout("latencyExperimentData");
  const latencyExperiment = latencyExperimentData.value;

  return (
    <WizardStep error={latencyExperimentData.error}>
      <MetadataTable
        onUpdate={(key, value) => latencyExperimentData.updateData(key, value)}
        data={[
          {
            name: "Percent",
            value: latencyExperiment.percent,
            input: {
              type: "number",
              key: "percent",
              validation: yup.number().integer().min(0).max(100),
            },
          },
          {
            name: "Duration (ms)",
            value: latencyExperiment.durationMs,
            input: {
              type: "number",
              key: "durationMs",
              validation: yup.number().integer().moreThan(0),
            },
          },
        ]}
      />
      <ButtonGroup
        buttons={[
          {
            text: "Back",
            onClick: onBack,
          },
          {
            text: "Next",
            onClick: onSubmit,
            destructive: true,
          },
        ]}
      />
    </WizardStep>
  );
};

const Confirm = () => {
  const startData = useDataLayout("startData");

  return (
    <WizardStep error={startData.error} isLoading={startData.isLoading}>
      <Confirmation action="Start" />
    </WizardStep>
  );
};

const createExperiment = data => {
  return client.post("/v1/experiments/create", {
    experiments: [
      {
        testSpecification: data,
      },
    ],
  });
};

export const StartAbortExperiment = ({ heading }) => {
  const dataLayout = {
    clusterPairTargetData: {},
    abortExperimentData: {},
    latencyExperimentData: {},
    startData: {
      deps: ["clusterPairTargetData", "abortExperimentData"],
      hydrator: (clusterPairTargetData, abortExperimentData) => {
        return createExperiment({
          abort: {
            clusterPair: {
              downstreamCluster: clusterPairTargetData.downstreamCluster,
              upstreamCluster: clusterPairTargetData.upstreamCluster,
            },
            percent: abortExperimentData.percent,
            httpStatus: abortExperimentData.httpStatus,
          },
        });
      },
    },
  };

  return (
    <Wizard dataLayout={dataLayout} heading={heading}>
      <ClusterPairTargetDetails name="target" />
      <AbortExperimentDetails name="abort" />
      <Confirm name="Confirmation" />
    </Wizard>
  );
};

export const StartLatencyExperiment = ({ heading }) => {
  const dataLayout = {
    clusterPairTargetData: {},
    latencyExperimentData: {},
    startData: {
      deps: ["clusterPairTargetData", "latencyExperimentData"],
      hydrator: (clusterPairTargetData, latencyExperimentData) => {
        return createExperiment({
          latency: {
            clusterPair: {
              downstreamCluster: clusterPairTargetData.downstreamCluster,
              upstreamCluster: clusterPairTargetData.upstreamCluster,
            },
            percent: latencyExperimentData.percent,
            durationMs: latencyExperimentData.durationMs,
          },
        });
      },
    },
  };

  return (
    <Wizard dataLayout={dataLayout} heading={heading}>
      <ClusterPairTargetDetails name="target" />
      <LatencyExperimentDetails name="latency" />
      <Confirm name="Confirmation" />
    </Wizard>
  );
};
