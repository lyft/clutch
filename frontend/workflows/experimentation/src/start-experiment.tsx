import React from "react";
import type { clutch as IClutch } from "@clutch-sh/api";
import type { BaseWorkflowProps } from "@clutch-sh/core";
import {
  ButtonGroup,
  client,
  Confirmation,
  MetadataTable,
  useWizardContext,
} from "@clutch-sh/core";
import { useDataLayout } from "@clutch-sh/data-layout";
import type { WizardChild } from "@clutch-sh/wizard";
import { Wizard, WizardStep } from "@clutch-sh/wizard";
import * as yup from "yup";

const ClusterPairTargetDetails: React.FC<WizardChild> = () => {
  const { onSubmit } = useWizardContext();
  const clusterPairData = useDataLayout("clusterPairTargetData");
  const clusterPair = clusterPairData.displayValue();

  return (
    <WizardStep error={clusterPairData.error} isLoading={false}>
      <MetadataTable
        onUpdate={(key, value) => clusterPairData.updateData(key, value)}
        data={[
          {
            name: "Downstream Cluster",
            value: clusterPair.downstreamCluster,
            input: {
              key: "downstreamCluster",
              validation: yup.string().required(),
            },
          },
          {
            name: "Upstream Cluster",
            value: clusterPair.upstreamCluster,
            input: {
              key: "upstreamCluster",
              validation: yup.string().required(),
            },
          },
        ]}
      />
      <ButtonGroup
        buttons={[
          {
            text: "Next",
            onClick: onSubmit,
          },
        ]}
      />
    </WizardStep>
  );
};

const AbortExperimentDetails: React.FC<WizardChild> = () => {
  const { onSubmit, onBack } = useWizardContext();
  const abortExperimentData = useDataLayout("abortExperimentData");
  const abortExperiment = abortExperimentData.value;

  return (
    <WizardStep error={abortExperimentData.error} isLoading={false}>
      <MetadataTable
        onUpdate={(key, value) => abortExperimentData.updateData(key, value)}
        data={[
          {
            name: "Percent",
            value: abortExperiment.percent,
            input: {
              type: "number",
              key: "percent",
              validation: yup.number().integer().moreThan(-1).lessThan(101),
            },
          },
          {
            name: "HTTP Status",
            value: abortExperiment.httpStatus,
            input: {
              type: "number",
              key: "httpStatus",
              validation: yup.number().integer().moreThan(99).lessThan(600),
            },
          },
        ]}
      />
      <ButtonGroup
        buttons={[
          {
            text: "Back",
            onClick: onBack,
          },
          {
            text: "Next",
            onClick: onSubmit,
            destructive: true,
          },
        ]}
      />
    </WizardStep>
  );
};

const LatencyExperimentDetails: React.FC<WizardChild> = () => {
  const { onSubmit, onBack } = useWizardContext();
  const latencyExperimentData = useDataLayout("latencyExperimentData");
  const latencyExperiment = latencyExperimentData.value;

  return (
    <WizardStep error={latencyExperimentData.error} isLoading={false}>
      <MetadataTable
        onUpdate={(key, value) => latencyExperimentData.updateData(key, value)}
        data={[
          {
            name: "Percent",
            value: latencyExperiment.percent,
            input: {
              type: "number",
              key: "percent",
              validation: yup.number().integer().min(0).max(100),
            },
          },
          {
            name: "Duration (ms)",
            value: latencyExperiment.durationMs,
            input: {
              type: "number",
              key: "durationMs",
              validation: yup.number().integer().moreThan(0),
            },
          },
        ]}
      />
      <ButtonGroup
        buttons={[
          {
            text: "Back",
            onClick: onBack,
          },
          {
            text: "Next",
            onClick: onSubmit,
            destructive: true,
          },
        ]}
      />
    </WizardStep>
  );
};

const Confirm: React.FC<WizardChild> = () => {
  const startData = useDataLayout("startData");

  return (
    <WizardStep error={startData.error} isLoading={startData.isLoading}>
      <Confirmation action="Start" />
    </WizardStep>
  );
};

const createExperiment = (data: IClutch.chaos.experimentation.v1.ITestSpecification) => {
  return client.post("/v1/experiments/create", {
    experiments: [
      {
        testSpecification: data,
      },
    ],
  });
};

export const StartAbortExperiment: React.FC<BaseWorkflowProps> = ({ heading }) => {
  const dataLayout = {
    clusterPairTargetData: {},
    abortExperimentData: {},
    latencyExperimentData: {},
    startData: {
      deps: ["clusterPairTargetData", "abortExperimentData"],
      hydrator: (
        clusterPairTargetData: IClutch.chaos.experimentation.v1.IClusterPairTarget,
        abortExperimentData: IClutch.chaos.experimentation.v1.AbortFault
      ) => {
        return createExperiment({
          abort: {
            clusterPair: {
              downstreamCluster: clusterPairTargetData.downstreamCluster,
              upstreamCluster: clusterPairTargetData.upstreamCluster,
            },
            percent: abortExperimentData.percent,
            httpStatus: abortExperimentData.httpStatus,
          },
        });
      },
    },
  };

  return (
    <Wizard dataLayout={dataLayout} heading={heading}>
      <ClusterPairTargetDetails name="target" />
      <AbortExperimentDetails name="abort" />
      <Confirm name="Confirmation" />
    </Wizard>
  );
};

export const StartLatencyExperiment: React.FC<BaseWorkflowProps> = ({ heading }) => {
  const dataLayout = {
    clusterPairTargetData: {},
    latencyExperimentData: {},
    startData: {
      deps: ["clusterPairTargetData", "latencyExperimentData"],
      hydrator: (
        clusterPairTargetData: IClutch.chaos.experimentation.v1.IClusterPairTarget,
        latencyExperimentData: IClutch.chaos.experimentation.v1.LatencyFault
      ) => {
        return createExperiment({
          latency: {
            clusterPair: {
              downstreamCluster: clusterPairTargetData.downstreamCluster,
              upstreamCluster: clusterPairTargetData.upstreamCluster,
            },
            percent: latencyExperimentData.percent,
            durationMs: latencyExperimentData.durationMs,
          },
        });
      },
    },
  };

  return (
    <Wizard dataLayout={dataLayout} heading={heading}>
      <ClusterPairTargetDetails name="target" />
      <LatencyExperimentDetails name="latency" />
      <Confirm name="Confirmation" />
    </Wizard>
  );
};
