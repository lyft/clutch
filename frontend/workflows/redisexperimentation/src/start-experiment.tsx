import React, { useState } from "react";
import { useForm } from "react-hook-form";
import type { clutch as IClutch } from "@clutch-sh/api";
import type { BaseWorkflowProps, ClutchError } from "@clutch-sh/core";
import {
  Button,
  ButtonGroup,
  client,
  Dialog,
  DialogActions,
  DialogContent,
  Form,
  useNavigate,
} from "@clutch-sh/core";
import { FormFields, PageLayout } from "@clutch-sh/experimentation";
import { yupResolver } from "@hookform/resolvers/yup";
import * as yup from "yup";

enum FaultType {
  ERROR = "Error",
  LATENCY = "Latency",
}

type ExperimentData = {
  downstreamCluster: string;
  upstreamRedisCluster: string;
  environmentValue: string;
  faultType: FaultType;
  percentage: number;
};

interface ExperimentDetailsProps {
  environments: Environment[];
  onStart: (ExperimentData) => void;
}

const ExperimentDetails: React.FC<ExperimentDetailsProps> = ({ environments, onStart }) => {
  const initialExperimentData = {
    faultType: FaultType.ERROR,
    environmentValue: environments.length > 0 ? environments[0].value : "",
  } as ExperimentData;

  const experimentDataState = useState<ExperimentData>(initialExperimentData);
  const experimentData = experimentDataState[0];
  const navigate = useNavigate();

  const handleOnCancel = () => {
    navigate("/experimentation/list");
  };

  const handleOnSubmit = () => {
    onStart(experimentData);
  };

  const fields = [
    {
      name: "downstreamCluster",
      label: "Downstream Cluster",
      type: "text",
      validation: yup.string().label("Downstream Cluster").required(),
      inputProps: { defaultValue: undefined },
    },
    {
      name: "upstreamRedisCluster",
      label: "Upstream Redis Cluster",
      type: "text",
      validation: yup.string().label("Upstream Redis Cluster").required(),
      inputProps: { defaultValue: undefined },
    },
    {
      name: "environmentValue",
      label: "Environment",
      type: "select",
      inputProps: {
        options: environments.map(env => {
          return {
            label: env.label,
            value: env.value,
          };
        }),
        defaultValue: initialExperimentData.environmentValue,
      },
      visible: environments.length > 0,
    },
    {
      name: "percentage",
      label: "Percent",
      type: "number",
      validation: yup.number().label("Percent").integer().min(1).max(100).required(),
      inputProps: { defaultValue: "0" },
    },
    {
      name: "faultType",
      label: "Fault Type",
      type: "select",
      inputProps: {
        options: [
          { label: "Error", value: FaultType.ERROR },
          { label: "Latency", value: FaultType.LATENCY },
        ],
        defaultValue: initialExperimentData.faultType,
      },
    },
  ];

  const schema: { [name: string]: yup.StringSchema | yup.NumberSchema } = {};
  fields
    .filter(field => field.validation !== undefined)
    .reduce((accumulator, field) => {
      if (field.validation) accumulator[field.name] = field.validation;
      return accumulator;
    }, schema);

  const { register, errors, handleSubmit } = useForm({
    mode: "onChange",
    reValidateMode: "onChange",
    resolver: yupResolver(yup.object().shape(schema)),
  });

  return (
    <Form onSubmit={handleSubmit(handleOnSubmit)}>
      <FormFields state={experimentDataState} items={fields} register={register} errors={errors} />
      <ButtonGroup>
        <Button text="Cancel" variant="neutral" onClick={handleOnCancel} />
        <Button text="Start" type="submit" />
      </ButtonGroup>
    </Form>
  );
};

type Environment = {
  // If provided, it's displayed to the user instead of environment's value i.e. 'Staging' or 'Production'.
  label?: string;
  // The value that represents the environment i.e. 'staging' or 'production'.
  value: string;
};

interface StartExperimentProps extends BaseWorkflowProps {
  // The template that's used to resolve the final name of an upstream redis cluster i.e., "$[CLUSTER]-$[ENVIRONMENT]".
  // It defaults to '$[CLUSTER]' and supports the following variables:
  //  1. $CLUSTER - the value that a user entered in upstream cluster input field.
  //  2. $ENVIRONMENT - the value of environment that a user selected. It's available only if
  //                    'environments' is provided. Resolves to an empty string otherwise.
  upstreamRedisClusterTemplate?: string;
  // The template that's used to resolve the final name of a downstream cluster i.e., "$[CLUSTER]-$[ENVIRONMENT]".
  // It defaults to '$[CLUSTER]' and supports the following variables:
  //  1. $CLUSTER - the value that a user entered in upstream cluster input field.
  //  2. $ENVIRONMENT - the value of environment that a user selected. It's available only if
  //                    'environments' is provided. Resolves to an empty string otherwise.
  downstreamClusterTemplate?: string;
  // The list of environments that a user should be able to choose from
  environments?: Environment[];
}

const StartExperiment: React.FC<StartExperimentProps> = ({
  heading,
  upstreamRedisClusterTemplate,
  downstreamClusterTemplate,
  environments,
}) => {
  const navigate = useNavigate();
  const [error, setError] = useState<ClutchError | undefined>(undefined);
  const [experimentData, setExperimentData] = useState<ExperimentData | undefined>(undefined);

  const handleOnCreatedExperiment = (id: number) => {
    navigate(`/experimentation/run/${id}`);
  };

  const handleOnCreatedExperimentFailure = (err: ClutchError) => {
    setExperimentData(undefined);
    setError(err);
  };

  const evaluateClusterTemplate = (
    clusterTemplate: string,
    cluster: string,
    environment: string
  ) => {
    return clusterTemplate.replace("$[CLUSTER]", cluster).replace("$[ENVIRONMENT]", environment);
  };

  const createExperiment = (data: ExperimentData) => {
    const faultTargeting = {
      upstreamCluster: {
        name: data?.upstreamRedisCluster,
      },
      downstreamCluster: {
        name: data?.downstreamCluster,
      },
    } as IClutch.chaos.redisexperimentation.v1.FaultTargeting;

    const isError = data?.faultType === FaultType.ERROR;
    const fault = isError
      ? {
          errorFault: {
            percentage: {
              percentage: data?.percentage,
            },
          },
        }
      : {
          latencyFault: {
            percentage: {
              percentage: data?.percentage,
            },
          },
        };

    const testConfig = {
      "@type": "type.googleapis.com/clutch.chaos.redisexperimentation.v1.FaultConfig",
      faultTargeting,
      ...fault,
    };

    return client
      .post("/v1/chaos/experimentation/createExperiment", {
        data: {
          config: testConfig,
        },
      })
      .then(response => {
        handleOnCreatedExperiment(response?.data.experiment.runId);
      })
      .catch((err: ClutchError) => {
        handleOnCreatedExperimentFailure(err);
      });
  };

  return (
    <PageLayout heading={heading} error={error}>
      <ExperimentDetails
        environments={environments ?? []}
        onStart={experimentDetails => setExperimentData(experimentDetails)}
      />
      <Dialog
        title="Experiment Start Confirmation"
        open={experimentData !== undefined}
        onClose={() => setExperimentData(undefined)}
      >
        <DialogContent>
          Are you sure you want to start an experiment? The experiment will start immediately and
          you will be moved to experiment details view page.
        </DialogContent>
        <DialogActions>
          <Button text="No" variant="neutral" onClick={() => setExperimentData(undefined)} />
          <Button
            text="Yes"
            onClick={() => {
              const environment = experimentData.environmentValue;
              experimentData.upstreamRedisCluster = evaluateClusterTemplate(
                upstreamRedisClusterTemplate ?? "$[CLUSTER]",
                experimentData.upstreamRedisCluster,
                environment
              );
              experimentData.downstreamCluster = evaluateClusterTemplate(
                downstreamClusterTemplate ?? "$[CLUSTER]",
                experimentData.downstreamCluster,
                environment
              );
              createExperiment(experimentData);
            }}
          />
        </DialogActions>
      </Dialog>
    </PageLayout>
  );
};

export { StartExperiment, ExperimentDetails };
