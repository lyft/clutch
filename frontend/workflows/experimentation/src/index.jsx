import ListExperiments from "./list-experiments";
import { StartAbortExperiment, StartLatencyExperiment } from "./start-experiment";

const register = function register() {
  return {
    developer: {
      name: "Lyft",
      contactUrl: "mailto:hello@clutch.sh",
    },
    path: "experimentation",
    group: "Experimentation",
    displayName: "Experimentation",
    routes: {
      listExperiments: {
        path: "list",
        displayName: "List Experiments",
        description: "List Experiments.",
        component: ListExperiments,
      },
      startAbortExperiment: {
        path: "startabort",
        displayName: "Start an Abort Experiment",
        description: "Start an Abort Experiment.",
        component: StartAbortExperiment,
      },
      startLatencyExperiment: {
        path: "startlatency",
        displayName: "Start a Latency Experiment",
        description: "Start a Latency Experiment.",
        component: StartLatencyExperiment,
      },
    },
  };
};

export default register;
