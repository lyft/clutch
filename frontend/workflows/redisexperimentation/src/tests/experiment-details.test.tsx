import React from "react";
import { FormFields } from "@clutch-sh/experimentation";
import { shallow } from "enzyme";

import { ExperimentDetails } from "../start-experiment";

jest.mock("@clutch-sh/core", () => {
  return {
    ...(jest.requireActual("@clutch-sh/core") as any),
    useNavigate: jest.fn(),
  };
});

describe("Start Experiment workflow", () => {
  it("renders correctly", () => {
    const component = shallow(<ExperimentDetails environments={[]} onStart={() => {}} />);
    expect(component.find(FormFields).dive().debug()).toMatchSnapshot();
  });

  it("renders correctly with upstream cluster type selection enabled", () => {
    const component = shallow(<ExperimentDetails environments={[]} onStart={() => {}} />);
    expect(component.find(FormFields).dive().debug()).toMatchSnapshot();
  });

  it("renders correctly with environments selection enabled", () => {
    const component = shallow(
      <ExperimentDetails
        downstreamClusterTemplate=""
        environments={[{ value: "staging" }]}
        onStart={() => {}}
      />
    );
    expect(component.find(FormFields).dive().debug()).toMatchSnapshot();
  });
});
