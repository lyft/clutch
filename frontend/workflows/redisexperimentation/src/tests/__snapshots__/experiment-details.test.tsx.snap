// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Start Experiment workflow renders correctly 1`] = `
"<Styled(div)>
  <TextField name=\\"downstreamCluster\\" label=\\"Downstream Cluster\\" defaultValue={[undefined]} type=\\"text\\" onChange={[Function: onChange]} inputRef={[Function: register]} error={false} helperText=\\"\\" />
  <TextField name=\\"upstreamRedisCluster\\" label=\\"Upstream Redis Cluster\\" defaultValue={[undefined]} type=\\"text\\" onChange={[Function: onChange]} inputRef={[Function: register]} error={false} helperText=\\"\\" />
  <Select name=\\"environmentValue\\" label=\\"Environment\\" options={{...}} defaultOption={-1} onChange={[Function: onChange]} />
  <TextField name=\\"percentage\\" label=\\"Percent\\" defaultValue=\\"0\\" type=\\"number\\" onChange={[Function: onChange]} inputRef={[Function: register]} error={false} helperText=\\"\\" />
  <Select name=\\"faultType\\" label=\\"Fault Type\\" options={{...}} defaultOption={0} onChange={[Function: onChange]} />
</Styled(div)>"
`;

exports[`Start Experiment workflow renders correctly with environments selection enabled 1`] = `
"<Styled(div)>
  <TextField name=\\"downstreamCluster\\" label=\\"Downstream Cluster\\" defaultValue={[undefined]} type=\\"text\\" onChange={[Function: onChange]} inputRef={[Function: register]} error={false} helperText=\\"\\" />
  <TextField name=\\"upstreamRedisCluster\\" label=\\"Upstream Redis Cluster\\" defaultValue={[undefined]} type=\\"text\\" onChange={[Function: onChange]} inputRef={[Function: register]} error={false} helperText=\\"\\" />
  <Select name=\\"environmentValue\\" label=\\"Environment\\" options={{...}} defaultOption={0} onChange={[Function: onChange]} />
  <TextField name=\\"percentage\\" label=\\"Percent\\" defaultValue=\\"0\\" type=\\"number\\" onChange={[Function: onChange]} inputRef={[Function: register]} error={false} helperText=\\"\\" />
  <Select name=\\"faultType\\" label=\\"Fault Type\\" options={{...}} defaultOption={0} onChange={[Function: onChange]} />
</Styled(div)>"
`;

exports[`Start Experiment workflow renders correctly with upstream cluster type selection enabled 1`] = `
"<Styled(div)>
  <TextField name=\\"downstreamCluster\\" label=\\"Downstream Cluster\\" defaultValue={[undefined]} type=\\"text\\" onChange={[Function: onChange]} inputRef={[Function: register]} error={false} helperText=\\"\\" />
  <TextField name=\\"upstreamRedisCluster\\" label=\\"Upstream Redis Cluster\\" defaultValue={[undefined]} type=\\"text\\" onChange={[Function: onChange]} inputRef={[Function: register]} error={false} helperText=\\"\\" />
  <Select name=\\"environmentValue\\" label=\\"Environment\\" options={{...}} defaultOption={-1} onChange={[Function: onChange]} />
  <TextField name=\\"percentage\\" label=\\"Percent\\" defaultValue=\\"0\\" type=\\"number\\" onChange={[Function: onChange]} inputRef={[Function: register]} error={false} helperText=\\"\\" />
  <Select name=\\"faultType\\" label=\\"Fault Type\\" options={{...}} defaultOption={0} onChange={[Function: onChange]} />
</Styled(div)>"
`;
